### ISSUE TITLE
DateTime passed into query producing incorrect results

### ISSUE THREAD

**Issue by tomtaylor** [elixir-sqlite/ecto_sqlite3#74]
I've got a table that looks like this:

```sql
CREATE TABLE IF NOT EXISTS "things" ("id" INTEGER PRIMARY KEY, "thing_id" INTEGER NOT NULL, "user_id" INTEGER NOT NULL, "timestamp" TEXT_DATETIME NOT NULL);
```

With data that looks like:

```
1|1|2|2016-03-09 23:26:46
2|2|3|2016-03-11 18:56:01
3|3|2|2016-03-08 15:24:38
4|4|2|2016-03-08 14:23:26
5|5|2|2016-03-08 10:02:39
6|6|2|2016-03-07 20:18:08
7|7|2|2016-03-02 15:30:22
8|8|2|2016-03-01 13:59:31
9|9|2|2016-02-25 19:02:16
10|10|2|2016-02-24 17:49:14
...
```

And a query that's very similar to this:

```sql
Thing
|> select([f], {min(f.timestamp), f.thing_id})
|> where([f], f.user_id in ^user_ids)
|> where([f], fragment("(?, ?) < (?, ?)", f.timestamp, f.thing_id, ^since_timestamp, ^since_thing_id))
|> group_by([f], f.thing_id)
|> order_by([f], desc: min(f.timestamp), desc: f.thing_id)
|> limit(^count)
```

I've been tearing my hair out trying to work out why it works fine in my SQLite console, but through Ecto it doesn't seem to produce the right results - the dates of the result are incorrect. I was passing a `DateTime` in as `since_timestamp`, but I've discovered the query only works correctly if I pass in the same thing in as an ISO8601 String. So `"2022-05-29 06:34:29"` instead of `~U[2022-05-29 06:34:29Z]`.

I guessing this is something to do with how a `DateTime` is cast into a string into Ecto, which is different to the string format in the database. It's really difficult to debug, because I can't see exactly what query is being produced.

Let me know if I can provide any more information or a test case.

**Comment by warmwaffles** [comment]
@tomtaylor also can you throw `IO.inspect(Ecto.Adapters.SQL.to_sql(:all, Repo, queryable))` into your query chain and post back the generated query?

**Comment by ruslandoga** [comment]
@warmwaffles I've created a repo for reproduction https://github.com/ruslandoga/issue74 and I'm not seeing `Ecto.Adapters.SQLite3.Codec.utc_datetime_encode` being called for datetime when using them in a fragment. I used rexbug for tracing.

**Comment by warmwaffles** [comment]
I'll take a look into this soon unless @ruslandoga you find a solution before I do.

**Comment by ruslandoga** [comment]
@warmwaffles I don't think there is any fix due to sqlite's lack of a read datetime type. As far as I can see, the best thing Ecto can do is cast the input, but SQLite seems to ignore it.

```sql
sqlite> SELECT * FROM things WHERE (timestamp < cast('2016-02-25T19:02:15' as text_datetime));
┌────┬──────────┬─────────┬─────────────────────┐
│ id │ thing_id │ user_id │      timestamp      │
├────┼──────────┼─────────┼─────────────────────┤
│ 9  │ 9        │ 2       │ 2016-02-25 19:02:16 │
│ 10 │ 10       │ 2       │ 2016-02-24 17:49:14 │
└────┴──────────┴─────────┴─────────────────────┘
```

**Comment by tomtaylor** [comment]
Thanks for everyone's work looking at this, and for providing a minimal test case.

I feels like the current behaviour is a little surprising. While SQLite has no native datetime format, the implication of the `datetime` function is that datetimes are best stored as `YYYY-MM-DD HH:MM:SS`, not `YYYY-MM-DDTHH:MM:SSZ`. 

Is this something we could make configurable? Perhaps defaulting to `YYYY-MM-DD HH:MM:SS` with ISO8601 configurable as an option? I imagine most databases use an internally consistent format, so making this configurable at the library level might be reasonable for most users.

**Comment by warmwaffles** [comment]
> Is this something we could make configurable?

A PR for the change would be welcome.

We simply defaulted to `DateTime.to_iso8601/1` for serializing to the database because it was simplest and most of sqlite's datetime https://www.sqlite.org/lang_datefunc.html expects the datetime to be in a subset of ISO-8601.

**Comment by greg-rychlewski** [comment]
@tomtaylor you can give the changes in this PR a try: https://github.com/elixir-sqlite/ecto_sqlite3/pull/84

**Comment by warmwaffles** [comment]
Lemme get a release cut

**Comment by warmwaffles** [comment]
The recent set of changes has been released under `v0.8.0`
